### PR TITLE
Python SDK dev module

### PR DESCRIPTION
**Pull request by helderco**
## Summary

This change adds a module for developing the Python SDK.

## Motivation

Main purpose is to allow others to run Python tasks without having to have a Python environment in their computers. This can be done in phases but includes something from all of the following:

1. Migrate hatch tasks;
2. Add utilities for testing or linting any Python files (e.g., doc snippets);
3. Migrate CI tasks (from mage);

The last one is related to a backlog item for using zenith in our CI. The `./ci` module will use `./sdk/python/dev` as a dependency.

## Interesting patterns

I found the following patterns interesting while making this module.

### Project structure

It’s a good example of how to structure bigger projects into multiple files. Instead of a single `src/main.py` file, you can turn it into a `src/main` package, with `__init__.py` as the “entrypoint” and other files imported from there.

### Constructors

With latest constructor support in https://github.com/dagger/dagger/pull/6152 it’s also possible to re-use an object’s constructor in another object’s method or top level function.

So the following lines allow breaking into dedicated objects various aspects of development, and provide top-level functions to instantiate those objects, without duplication:

https://github.com/dagger/dagger/blob/3f0163be8d8565e449ce8c7bb0903e0b0802f435/sdk/python/dev/src/main/__init__.py#L26-L30

### Laziness

The object types were created with modularity principles in mind, and maximum laziness. Especially in the `Deps` object, there’s no special care to resolve things before using. By passing around lazy objects like any programing object, the engine resolves the DAG more efficiently.

https://github.com/dagger/dagger/blob/3f0163be8d8565e449ce8c7bb0903e0b0802f435/sdk/python/dev/src/main/deps.py#L219-L225

I was surprised however at how little async calls I actually needed in the end seeing how much this module actually does. One is a joke[^1], the other is an experiment and the last one was needed to get the an `endpoint` value for a test condition (`Test.provision`) that’s hardcoded right now in  a GHA workflow that’s only run when PR’s are merged into main. I tested the condition locally on macOS and it failed in Linux on CI. This test reproduces that environment with docker so anyone can test it locally.

[^1]: `dagger -m dev call joke`: "['hip', 'hip'] (hip hip array!)"

### Concurrency

The experiment is between `dagger call all-tests` (lazy) and `dagger-call tests` (concurrent). One uses concurrency to test the SDK with multiple Python versions, while the other just returns a list of lazy `Test` objects for each version. 

https://github.com/dagger/dagger/blob/3f0163be8d8565e449ce8c7bb0903e0b0802f435/sdk/python/dev/src/main/__init__.py#L98

What’s interesting is that returning a list of functions seems viable to replace the need for concurrency at the language level in some cases.

If you run `dagger call tests` it’s fixed to executing `Test.default`, which runs all tests. If you just want to run unit tests you’re out of luck if you want a single command. But with `dagger call all-tests unit`, you’re basically selecting a field for each item the the list so there’s more freedom to sub-select a function on multiple objects, while executing them concurrently and without having to do any concurrency at the language level.

### `with_focus` and `pipeline`

If you rely on `dagger call` adding `.sdout()` to a `Container`, then you’ll need the let it finish to see the test run’s results, unless you have `--focus=false`. But if you take care in adding `.with_focus()` before the execs that you’re interesting in seeing the output of, then it’ll stream that output in the TUI during execution in the default `--focus=true`, without showing the clutter.

https://github.com/dagger/dagger/blob/3f0163be8d8565e449ce8c7bb0903e0b0802f435/sdk/python/dev/src/main/test.py#L56-L61

This is especially nice when you’re returning multiple lazy `Test` objects for different Python versions and seeing their output in real-time, concurrently.

There’s a catch 22 problem however. On one hand, the automatic addition of `.stdout` by `dagger call` creates duplicated output (one from the *focused* step, the other from the final response). On the other, if you don’t add `stdout` you wouldn’t see anything without the TUI (e.g., `--silent`).

There’s several ways this could be fixed but I think relying more on `with_focus` provides the better experience. Paired with `pipeline` you can group the output in the TUI with your own labels, while with `stdout` from multiple objects you get all the output but they’re not categorized or grouped. Just printed one after the other, in the ordered the happened to arrive.

## What’s next?

- Directories are only filtered when using the default sources (from `host`). If you provide an alternative dir via the CLI, that’s not filtered. Try to exclude files from any `Directory` instead.
- Port more functions from mage and use as dependency of a `./ci` module in Go, to delegate all Python related tasks to, for example, linting doc snippets. Related to:
    - https://github.com/dagger/dagger/issues/6137
- Document how to:
    - test a doc snippet using the module
    - lint doc snippets

### CODE REVIEW

**Comment by helderco** [comment]
Ready for review. Updated the description with more context.

**Comment by helderco** [comment]
Rebased.

Note for reviewer. While there's some functions that don't work quite right from outside of the `sdk/phython` dir, like linting docs, the module does what was intended initially: to replace _hatch_ tasks with dagger. CI with mage works as usual, replacing it is part of a larger effort and out of scope for this PR.

Easier to make improvements over time in followups because while this isn't merged it takes effort to use it in other branches.

**Comment by gerhard** [comment]
Starting to look into this one now.

**Comment by helderco** [comment]
@gerhard you may need `hack/dev` if you haven't built it. It depends on _main_.

File: sdk/python/dev/src/main/test.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+            .as_service()
+        )
+
+        http_server_url = await http_server.endpoint(scheme="http")
+        archive_url = f"{http_server_url}/{archive_name}"
+        checksums_url = f"{http_server_url}/{checksums_name}"
+
+        docker_version = "24.0.7"
+
+        ctr = dag.dockerd().attach(
```
Comment (sipsma) [inline sdk/python/dev/src/main/test.py:121]: Nice! Can't wait to migrate all the provisioning tests to this approach 😃

File: sdk/python/dev/src/main/test.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+            str | None,
+            Doc("_EXPERIMENTAL_DAGGER_RUNNER_HOST value"),
+        ] = None,
+    ) -> dagger.Container:
+        """Test provisioning.
+
+        This publishes a cli binary in an ephemeral http server and checks
+        if the SDK can download, extract and run it.
+        """
+        # TODO: Most of this setup can be reused for all SDKs that need to
```
Comment (sipsma) [inline sdk/python/dev/src/main/test.py:92]: 😎

**Review by sipsma — APPROVED** [review]
This is really really really cool, so much neat stuff. The idea of returning a list of `Test` objects and then selecting a container from them to run a specific subtest is super interesting, seems like a very re-usable general idea!

---
The only thing I noticed is that I'm currently getting this when I run `../../hack/dev dagger call -m dev tests` 

```
+ exec dagger call -m dev tests
✔ dagger call tests [48.49s]
┃ true                                                                                                                                                                                                                                                  
✘ sync ERROR [0.01s]
✘ exec foobar ERROR [0.13s]
┃ panic: exec: "foobar": executable file not found in $PATH goroutine 1 [running, locked to thread]:                                                                                                                                                    
┃ runtime/debug.Stack()                                                                                                                                                                                                                                 
┃         /usr/local/go/src/runtime/debug/stack.go:24 +0x64                                                                                                                                                                                             
┃ main.main.func1()                                                                                                                                                                                                                                     
┃         /app/cmd/shim/main.go:63 +0x34                                                                                                                                                                                                                
┃ panic({0xa7dae0?, 0x40005d8b80?})                                                                                                                                                                                                                     
┃         /usr/local/go/src/runtime/panic.go:914 +0x218                                                                                                                                                                                                 
┃ main.shim()                                                                                                                                                                                                                                           
┃         /app/cmd/shim/main.go:291 +0xda8                                                                                                                                                                                                              
┃ main.main()                                                                                                                                                                                                                                           
┃         /app/cmd/shim/main.go:76 +0x98                                                                                                                                                                                                                
✘ directory(path: "foobar") ERROR [0.02s]
```

However, quite mysteriously I hit that, had to go look at something else for a few hours and then when I came back and ran it again it passed. So could be a flake? Just worth double checking

**Comment by helderco** [comment]
> The only thing I noticed is that I'm currently getting this when I run `../../hack/dev dagger call -m dev tests`
> 
> ```
> + exec dagger call -m dev tests
> ✔ dagger call tests [48.49s]
> ┃ true                                                                                                                                                                                                                                                  
> ✘ sync ERROR [0.01s]
> ✘ exec foobar ERROR [0.13s]
> ┃ panic: exec: "foobar": executable file not found in $PATH goroutine 1 [running, locked to thread]:                                                                                                                                                    
> ┃ runtime/debug.Stack()                                                                                                                                                                                                                                 
> ┃         /usr/local/go/src/runtime/debug/stack.go:24 +0x64                                                                                                                                                                                             
> ┃ main.main.func1()                                                                                                                                                                                                                                     
> ┃         /app/cmd/shim/main.go:63 +0x34                                                                                                                                                                                                                
> ┃ panic({0xa7dae0?, 0x40005d8b80?})                                                                                                                                                                                                                     
> ┃         /usr/local/go/src/runtime/panic.go:914 +0x218                                                                                                                                                                                                 
> ┃ main.shim()                                                                                                                                                                                                                                           
> ┃         /app/cmd/shim/main.go:291 +0xda8                                                                                                                                                                                                              
> ┃ main.main()                                                                                                                                                                                                                                           
> ┃         /app/cmd/shim/main.go:76 +0x98                                                                                                                                                                                                                
> ✘ directory(path: "foobar") ERROR [0.02s]
> ```
> 
> However, quite mysteriously I hit that, had to go look at something else for a few hours and then when I came back and ran it again it passed. So could be a flake? Just worth double checking

That's a known issue actually! **It's harmless** but we should fix it I think. I forgot to create an issue.

There's a few integration tests around `.sync` that trigger errors on purpose because `.sync` behavior is harder to test on a successful case.

<img width="648" alt="Screenshot 2023-12-13 at 10 09 44" src="https://github.com/dagger/dagger/assets/174525/0a436bff-4934-4837-ad53-bd459a68c118">

@sipsma So the Go panic is being raised on non-existing `withExec` command and `directory` path. I haven't looked into `shim/main.go` yet.

**Pull request merged**